File: output/GPT4o-Mini/xCodeEval_remove_4000/Java/Go/26a05f9a232a71950fde9cd7134ffd34.go
package main

import (
	"bufio"
	"fmt"
	"math/big"
	"os"
	"sort"
)

const MODULO = 1000000007

type Pair struct {
	power, coeff int64
}

var id34 int64

func main() {
	r := bufio.NewReader(os.Stdin)
	w := bufio.NewWriter(os.Stdout)
	defer w.Flush()

	n, s := readInt(r), readInt64(r)

	f := id12(n)

	coeffs := make(map[int64]int64)
	coeffs[0] = 1

	for _, size := range f {
		coeffEntries := []Pair{}
		for power, coeff := range coeffs {
			nextPower := power + size + 1
			coeffEntries = append(coeffEntries, Pair{nextPower, -coeff})
		}

		for _, coeffEntry := range coeffEntries {
			coeffs[coeffEntry.power] = add(coeffs[coeffEntry.power], coeffEntry.coeff)
		}
	}

	id25 := multSequence(1, int64(n-1))
	id34 = inverse(id25)

	answer := int64(0)
	for power, coeff := range coeffs {
		if s >= power {
			need := s - power
			answer = add(answer, mult(binom(int64(n)+need-1, need), coeff))
		}
	}

	fmt.Fprintln(w, answer)
}

func multSequence(start, end int64) int64 {
	res := int64(1)
	for i := start; i <= end; i++ {
		res = mult(res, i)
	}
	return res
}

func inverse(x int64) int64 {
	return binpow(x, MODULO-2)
}

func binpow(base, power int64) int64 {
	if power == 0 {
		return 1
	}
	if power%2 == 0 {
		half := binpow(base, power/2)
		return mult(half, half)
	}
	return mult(binpow(base, power-1), base)
}

func binom(n, k int64) int64 {
	return mult(multSequence(k+1, n), id34)
}

func add(a, b int64) int64 {
	return ((a%MODULO + b%MODULO) % MODULO + MODULO) % MODULO
}

func mult(a, b int64) int64 {
	return (a % MODULO) * (b % MODULO) % MODULO
}

func readInt(r *bufio.Reader) int {
	var x int
	fmt.Fscan(r, &x)
	return x
}

func readInt64(r *bufio.Reader) int64 {
	var x int64
	fmt.Fscan(r, &x)
	return x
}

func id12(size int) []int64 {
	array := make([]int64, size)
	for i := 0; i < size; i++ {
		array[i] = readInt64(bufio.NewReader(os.Stdin))
	}
	return array
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/Java/C/26a05f9a232a71950fde9cd7134ffd34.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>
#include <string.h>

#define MAX_SIZE 1000000
#define MODULO 1000000007

long long id34;

long long add(long long a, long long b) {
    return ((a % MODULO + b % MODULO) % MODULO + MODULO) % MODULO;
}

long long mult(long long a, long long b) {
    return (a % MODULO) * (b % MODULO) % MODULO;
}

long long binpow(long long base, long long power) {
    if (power == 0) return 1;

    if ((power & 1) == 0) {
        long long half = binpow(base, power >> 1);
        return mult(half, half);
    } else {
        long long prev = binpow(base, power - 1);
        return mult(prev, base);
    }
}

long long inverse(long long x) {
    return binpow(x, MODULO - 2);
}

long long multSequence(long long start, long long end) {
    long long res = 1;
    for (long long i = start; i <= end; ++i) {
        res = mult(res, i);
    }

    return res;
}

long long binom(long long n, long long k) {
    return mult(multSequence(k + 1, n), id34);
}

void solve() {
    int n;
    long long s;

    scanf("%d %lld", &n, &s);

    long long f[MAX_SIZE];
    for (int i = 0; i < n; i++) {
        scanf("%lld", &f[i]);
    }

    long long coeffs[MAX_SIZE];
    memset(coeffs, 0, sizeof(coeffs));
    coeffs[0] = 1;

    for (int i = 0; i < n; i++) {
        long long size = f[i];
        long long coeffEntries[MAX_SIZE][2];
        int coeffEntriesCount = 0;

        for (int j = 0; j < MAX_SIZE; j++) {
            if (coeffs[j] != 0) {
                long long power = j;
                long long coeff = coeffs[j];

                long long nextPower = power + size + 1;
                coeffEntries[coeffEntriesCount][0] = nextPower;
                coeffEntries[coeffEntriesCount][1] = -coeff;
                coeffEntriesCount++;
            }
        }

        for (int j = 0; j < coeffEntriesCount; j++) {
            long long power = coeffEntries[j][0];
            long long coeff = coeffEntries[j][1];

            coeffs[power] = add(coeffs[power], coeff);
        }
    }

    long long id25 = multSequence(1, n - 1);
    id34 = inverse(id25);

    long long answer = 0;
    for (int i = 0; i < MAX_SIZE; i++) {
        if (coeffs[i] != 0) {
            long long power = i;
            long long coeff = coeffs[i];

            if (s >= power) {
                long long need = s - power;
                answer = add(answer, mult(binom(n + need - 1, need), coeff));
            }
        }
    }

    printf("%lld\n", answer);
}

int main() {
    solve();

    return 0;
}
